### ISSUE TITLE
Simplify subscriptions.yaml

### ISSUE THREAD

**Issue by jmbannon** [jmbannon/ytdl-sub#743]
An extension of https://ytdl-sub.readthedocs.io/en/latest/config.html#file-preset

Have the ability to do something like:
```
# applies to entire file
 __preset__:
   preset: "playlist_preset_ex"
__value__: "url"

# applies for subsequent URLs
__jazz__:
  overrides:
    genre: "Jazz"

jazz_artist1: "url"
jazz_artist2: "url"

# NEW sub-preset, discard __jazz__ and apply __rock__
__rock__:
  overrides:
    genre: "Rock"

rock_artist1: "url"
rock_artist2: "url"
```

**Comment by jmbannon** [comment]
Alternative idea: have the presets reside in the config.yaml, then do:
```
jazz:
  jazz_artist1: "url"
  jazz_artist2: "url"

rock:
  rock_artist1: "url"
  rock_artist2: "url"
```

Also could add support for nested presets:
```
jazz:
  recent_vids_only:
    jazz_artist1: "url"
  jazz_artist2: "url"

rock:
  rock_artist1: "url"
  rock_artist2: "url"
```

**Comment by Behem07h** [comment]
The text editor I use (notepad++) has collapsible indents, which helps make my subscriptions.yaml much more readable when i'm trying to edit it. Because of this I personally would much prefer
`
__preset__:
    preset: "multiurl_music_preset"

jazz:
    jazz_artist_1: 
        download:
            - "url 1"
            -  "url 2"
            
rock:
    rock_artist_1: 
        download:
            - "url 1"
            -  "url 2"
`

where "jazz" and "rock" are variables we can reference, just like {subscription_name} is. This is quite similar to your second proposed format, but it avoids the potential "preset bloat" of needing to have a ton of presets that only really change one or two variables. 

I do like your 1st proposed format, but I think it would be better used for compacting 2 subscription files that utilise 2 different presets, into 1 file. If used solely for the purpose of defining a genre override variable i think #.1 is too clunky

**Comment by psyciknz** [comment]
I kind like this one for it's simplicity:

```
jazz:  # applies jazz preset
  jazz_artist1: "url"
  jazz_artist2: "url"

rock:
  rock_artist1: "url"
  rock_artist2: "url"
```

Only because it gets a subscription to a single line.....which I guess is all a simpleton like me wants.    Even if the overally structure is complicated, then reducing a single subscription to a line that is easily construction while model (either in a terminal app on IOS or some really simple text editor - imo should be the goal.

**Comment by jmbannon** [comment]
I think I'm definitely set on being able to nest presets, then have the last key/value be the subscription name + subscription value (which maps to an override variable)

re @Behem07h 's idea of being able to set override variables without the need for a preset. I do like that and it would especially be useful for genres. What about this

```
youtube_discography:  # applies youtube_discography preset
  genre@Jazz:  # the @ signifies to set everything to the right of it into the `genre` override
    jazz_artist1: "url"
  "genre@Japanese Jazz": 
    japanese_jazz_artist: "url"
```

**Comment by jmbannon** [comment]
Alternative syntax for the above - not really liking the `@`
```
youtube_discography: 
  genre = Jazz:
    jazz_artist1: "url"
  genre = Japanese Jazz:
    japanese_jazz_artist: "url"
```

**Comment by Behem07h** [comment]
Another option I just came up with. personally i think this is the cleanest.

```
__preset__:
    preset: "music_preset"
    overrides:
        genre: "{indent1}"

Jazz:
    jazz_artist_1: "url"
    jazz_artist_2:
        - "url1"
        - "url2"
Rock:
    rock_artist_1: "url"


```

**Comment by jmbannon** [comment]
@Behem07h you need a way to tell which list items map to which override variables. The backend would have to assume a list means you want it underneath `download`, which is too specific

**Comment by jmbannon** [comment]
I'm hoping to add list support for overrides at some point, perhaps that's can be how it's done in the future

**Comment by jmbannon** [comment]
Regarding `indent1`, I like the idea, but would also like to have the ability to 'hide' the config layer when writing subscriptions. The idea being that 'prebuilt configs' could be distributed and users require zero knowledge of ytdl-sub aside from "put your values in this subscriptions file like this".

**Comment by Behem07h** [comment]
> Regarding `indent1`, I like the idea, but would also like to have the ability to 'hide' the config layer when writing subscriptions. The idea being that 'prebuilt configs' could be distributed and users require zero knowledge of ytdl-sub aside from "put your values in this subscriptions file like this".

In essence my idea was that just like {subscription_name}, we give every indentation its own variable name. for example
{indent_1}:
    {indent_2}:
        {indent_3}:
        etc...
            {subscription_name}:
                - "url"

This way people can use as many indents as they want, and use the variables in their config just like any variable. If the indent didnt exist in the subscription entry it could just return blank.

I showed {indent_1} being assigned to an override variable simply for illustration purposes to show how it all would work

**Comment by jmbannon** [comment]
Indent could work but you'd have to assign the override variable names ahead of time, which I think would look clunky

**Comment by Behem07h** [comment]
ah, i see what you mean now regarding the preset vs indent issue. indent2 vs subscription name isn't an issue because subscription name is always the last indent. 
I just think genre = Jazz: is too wordy, the subscription file wouldn't read well. Perhaps we reserve some character(s) to wrap preset names or indents in? Sort of thinking along the lines of how python differentiates between "strings" and f"fstrings". Perhaps we can just unallow [] or smth from being part of a preset name and use that to wrap indents? Unlike subscription names and indents, preset names are mainly descriptive and unlikely to contain special characters

**Comment by jmbannon** [comment]
> subscription name is always the last indent

It's not though when you define an actual preset in the last key, like my example above, where `download` is actually part of the preset

**Comment by jmbannon** [comment]
I agree `genre = Jazz:` is kinda weird to have an `=` assignment in a key.

**Comment by Behem07h** [comment]
maybe we can have the user define the indent structure in the config? i'm not quite sure how difficult that would be to parse and implement into code, but that would give the most flexibility, allow us to use arbitrary indent names without ambiguity, and keeps the subscription file clean looking.

One caveat to this would be potential confusion of indent formats between presets though

**Comment by jmbannon** [comment]
> One caveat to this would be potential confusion of indent formats between presets though

Right, this is my main concern. Another is, maybe you want to set multiple overrides but not have to nest so deep

**Comment by Behem07h** [comment]
> > subscription name is always the last indent
> 
> It's not though when you define an actual preset in the last key, like my example above, where `download` is actually part of the preset

i see what u mean about this now. Perhaps we can solve this by enforcing the amount of indents to be the amount the user uses in their config? e.g if the preset uses {indent1} and {indent2}, the third indent must be the subscription name

**Comment by jmbannon** [comment]
> Perhaps we can solve this by enforcing the amount of indents to be the amount the user uses in their config?

Right, like I said, it'd have to be predefined ahead either in the config or subscription, mapping indents to override variables. That could be done but again, I'm not a fan of the config 'locking' the format

**Comment by jmbannon** [comment]
+ the indent issue has a workaround by defining more presets in your config (one for each genre). It's cumbersome, but at least solves the 'pretty subscription file'

**Comment by Behem07h** [comment]
considering all this i think the best option is that we stick with the {indent _1} format which lets us use infinitely many indents, but we add an identifier to how the indent must be typed out to be recognised as one. 
Since we already have a predetermined variable name for each indent, we dont need to define the variable name like "genre = jazz:".

**Comment by jmbannon** [comment]
If we do that, we need some kind of identifier for presets, like
```
+youtube_discography:
  ...
```

**Comment by Behem07h** [comment]
> plus the indent issue has a workaround by defining more presets in your config (one for each genre). It's cumbersome, but at least solves the 'pretty subscription file'

that's true, but i figured we should stay away from making the config file messy like that. It adds extra work when adding a new genre to the subscription file, and it reminds me of the "url1, url2, url3" situation we had before multiurl download was a thing (i.e, something that could get replaced/superceded in the future due to its clunkiness)

**Comment by jmbannon** [comment]
I would prefer having indent variables to have the identifier over presets

**Comment by jmbannon** [comment]
Another thing that would allow us to do is assign indent variables not actually based on number of indents, but just number of indent variables, i.e.

```
tv_show:  # preset
  __TV-Y:  # first indent in this nest, set indent1=TV-Y
    name: url

  only_recent:
    __TV-Y:  # first indent in this nest, set indent1=TV-y
    other_name:  other_url
```

**Comment by jmbannon** [comment]
I'm actually really liking the syntax above. Just need to settle on a nice looking identifier

**Comment by jmbannon** [comment]
Here are my proposals so far for indent syntax:
```
tv_show: 
  __TV-Y: 
    name: url
```

```
tv_show: 
  =TV-Y: 
    name: url
```

```
tv_show: 
  [TV-Y]:   # this one may collide with future list support
    name: url
```

```
tv_show: 
  ~TV-Y: 
    name: url
```

I think I like `=` the best so far

**Comment by jmbannon** [comment]
More syntax ideas:
```
tv_show: 
  :TV-Y: 
    name: url
```

```
tv_show: 
  TV-Y:: 
    name: url
```

**Comment by jmbannon** [comment]
From the discord strawpoll, `[TV-Y]` was the clear winner. Will go with that

**Comment by Yankees4life** [comment]
How about having names be automatically assigned by ytdlsub? I got a lot of urls so setting names for each one can be tedious

**Comment by jmbannon** [comment]
> How about having sub names be automatically named by ytdlsub? I got a lot of urls so setting names for each one can be tedious

You could do this by setting the subscription_name to the url, and have an emtpy value. I.e.

```
tv_show:
  "https://youtube.com/@test":
  "https://youtube.com/@test2":
```
Then set the override variable `url: "{subscription_name}"`. You would need to set some other stuff up as well so the output directory doesn't include the tv_show_name (the prebuilt presets do this today). A few users do this today. Maybe we can make a separate feature request to make a prebuilt preset for this.

My personal recommendation is it's good to have an explicit name in case the channel changes its name. If they do, then videos will start saving into a different directory

**Comment by jmbannon** [comment]
Closing: https://ytdl-sub.readthedocs.io/en/latest/config.html#beautifying-subscriptions
For migration, see https://github.com/jmbannon/ytdl-sub/releases/tag/2023.10.17
